File: clients/python/guardian_client.py
"""core-guardian client — vendor this single file into any Python service.

Zero external dependencies (urllib only). Identity comes from the ``GUARDIAN``
config (a JSON string env var or a dict); the two token audiences come from the
``GUARDIAN_AI_TOKEN`` and ``GUARDIAN_API_KEY`` env vars. Mirrors the TypeScript
client's contract exactly. Source of truth:
https://github.com/jmbish04/core-guardian/blob/main/clients/python/guardian_client.py
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Callable, Dict, Iterator, Mapping, Optional, Tuple

VERSION = "1.0.0"
DEFAULT_BASE_URL = "https://core-guardian.hacolby.workers.dev"
PRIORITY_TO_IMPORTANCE: Dict[str, str] = {
    "hobby": "low",
    "normal": "low",
    "important": "medium",
    "critical": "high",
}

# An injected HTTP transport: (method, url, headers, body_bytes, stream) ->
# (status:int, body:bytes|Iterator[bytes]). Lets the self-check run without a
# network. Defaults to urllib.
Http = Callable[[str, str, Dict[str, str], Optional[bytes], bool], Tuple[int, Any]]


class GuardianError(Exception):
    """Raised for any non-2xx response. Mirrors the TS GuardianError."""

    def __init__(self, status: int, body: Any):
        super().__init__(f"Guardian request failed ({status})")
        self.status = status
        self.body = body
        b = body if isinstance(body, dict) else {}
        self.is_circuit_breaker = bool(b.get("isCircuitBreaker"))
        msg = b.get("circuitBrokenMessage")
        self.circuit_broken_message = msg if isinstance(msg, str) else None


def _drop_none(d: Dict[str, Any]) -> Dict[str, Any]:
    return {k: v for k, v in d.items() if v is not None}


def _parse_json(raw: Optional[bytes]) -> Any:
    # Catch ValueError (covers json.JSONDecodeError AND the UnicodeDecodeError a
    # binary/gzip error page would raise) so a bad body degrades to None, never
    # escaping past the caller's GuardianError handling.
    try:
        return json.loads(raw) if raw else None
    except ValueError:
        return None


def _urllib_http(
    method: str, url: str, headers: Dict[str, str], body: Optional[bytes], stream: bool, timeout: Optional[float] = 30.0
) -> Tuple[int, Any]:
    req = urllib.request.Request(url, data=body, method=method, headers=headers)
    try:
        # Streaming reads happen after urlopen returns, so the connect timeout
        # still applies but a long stream isn't cut off by it.
        resp = urllib.request.urlopen(req, timeout=timeout)  # noqa: S310 (fixed base URL, not user-controlled scheme)
    except urllib.error.HTTPError as e:
        raise GuardianError(e.code, _parse_json(e.read()))
    if stream:
        def chunks() -> Iterator[bytes]:
            while True:
                chunk = resp.read(8192)
                if not chunk:
                    break
                yield chunk
        return resp.status, chunks()
    return resp.status, _parse_json(resp.read())


class _AI:
    def __init__(self, client: "GuardianClient"):
        self._c = client

    def run(
        self,
        *,
        provider: str,
        model: str,
        input: Any,
        importance: Optional[str] = None,
        mode: Optional[str] = None,
        ai_gateway_id: Optional[str] = None,
        transport: Optional[str] = None,
        provider_api_key: Optional[str] = None,
    ) -> Any:
        return self._c._run(
            provider, model, input, importance, mode, ai_gateway_id, transport, provider_api_key, False
        )

    def stream(
        self,
        *,
        provider: str,
        model: str,
        input: Any,
        importance: Optional[str] = None,
        mode: Optional[str] = None,
        ai_gateway_id: Optional[str] = None,
        transport: Optional[str] = None,
        provider_api_key: Optional[str] = None,
    ) -> Iterator[bytes]:
        return self._c._run(
            provider, model, input, importance, mode, ai_gateway_id, transport, provider_api_key, True
        )


class _Usage:
    def __init__(self, client: "GuardianClient"):
        self._c = client

    def register(
        self,
        *,
        provider: str,
        model: str,
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
        tokens_thinking: Optional[int] = None,
        requests: Optional[int] = None,
        cost_usd: Optional[float] = None,
        operation_id: Optional[str] = None,
        task_description: Optional[str] = None,
    ) -> Any:
        return self._c._register(
            _drop_none({
                "provider": provider,
                "model": model,
                "tokensIn": tokens_in,
                "tokensOut": tokens_out,
                "tokensThinking": tokens_thinking,
                "requests": requests,
                "costUsd": cost_usd,
                "operationId": operation_id,
                "taskDescription": task_description,
            })
        )


class GuardianClient:
    VERSION = VERSION

    def __init__(
        self,
        *,
        project: str,
        repo: Optional[str] = None,
        priority: Optional[str] = None,
        budget: Optional[float] = None,
        base_url: Optional[str] = None,
        ai_token: Optional[str] = None,
        api_key: Optional[str] = None,
        http: Optional[Http] = None,
        timeout: Optional[float] = 30.0,
    ):
        if not project:
            raise ValueError("GuardianClient: project is required")
        self.project = project
        self.repo = repo
        self.priority = priority
        # Named budget_usd so it can't shadow the budget() endpoint method.
        self.budget_usd = budget
        self.base_url = (base_url or DEFAULT_BASE_URL).rstrip("/")
        # Leading underscore keeps the tokens out of a casual vars()/repr dump.
        self._ai_token = ai_token
        self._api_key = api_key
        # Default transport carries the timeout; an injected http keeps the
        # 5-arg signature (timeout is the transport's concern).
        self._http = http or (lambda m, u, h, b, s: _urllib_http(m, u, h, b, s, timeout))
        self.ai = _AI(self)
        self.usage = _Usage(self)

    @classmethod
    def from_env(cls, env: Optional[Mapping[str, Any]] = None, **kwargs: Any) -> "GuardianClient":
        env = env if env is not None else os.environ
        raw = env.get("GUARDIAN")
        if raw is None:
            raise ValueError("GuardianClient.from_env: GUARDIAN missing")
        try:
            cfg = json.loads(raw) if isinstance(raw, str) else dict(raw)
        except (ValueError, TypeError):
            raise ValueError("GuardianClient.from_env: GUARDIAN is not valid JSON or a mapping")
        if not cfg.get("project"):
            raise ValueError("GuardianClient.from_env: GUARDIAN.project missing")
        return cls(
            project=cfg["project"],
            repo=cfg.get("repo"),
            priority=cfg.get("priority"),
            budget=cfg.get("budget"),
            base_url=cfg.get("baseUrl"),
            ai_token=env.get("GUARDIAN_AI_TOKEN"),
            api_key=env.get("GUARDIAN_API_KEY"),
            **kwargs,
        )

    def _importance_for(self, over: Optional[str]) -> str:
        return over or PRIORITY_TO_IMPORTANCE.get(self.priority or "normal", "low")

    def _send(self, method: str, path: str, token: Optional[str], body: Optional[dict], stream: bool = False) -> Tuple[int, Any]:
        if not token:
            raise RuntimeError(f"GuardianClient: missing token for {path}")
        headers = {"authorization": f"Bearer {token}"}
        data: Optional[bytes] = None
        if body is not None:
            headers["content-type"] = "application/json"
            data = json.dumps(body).encode("utf-8")
        status, payload = self._http(method, self.base_url + path, headers, data, stream)
        # Raise on error for streaming too (matches the TS client): a 400 for a
        # non-openai stream or a 429 breaker must surface, not be iterated.
        if status >= 400:
            raise GuardianError(status, payload)
        return status, payload

    def _run(self, provider, model, input, importance, mode, ai_gateway_id, transport, provider_api_key, stream):
        body = _drop_none({
            "project": self.project,
            "importance": self._importance_for(importance),
            "provider": provider,
            "model": model,
            "input": input,
            "mode": mode,
            "aiGatewayId": ai_gateway_id,
            "transport": transport,
            "providerApiKey": provider_api_key,
            "stream": stream,
        })
        _status, payload = self._send("POST", "/api/ai-router/run", self._ai_token, body, stream)
        return payload

    def _register(self, u: Dict[str, Any]) -> Any:
        body = {"worker": self.project, **u}
        _status, payload = self._send("POST", "/api/guardian/usage/register", self._api_key, body)
        return payload

    def budget(self) -> Any:
        _status, payload = self._send("GET", "/api/ai/budget", self._api_key, None)
        return payload

    def project_status(self) -> Any:
        """GET the registered project record. (Named project_status to avoid
        shadowing the ``project`` identity attribute.)"""
        path = "/api/guardian/projects/" + urllib.parse.quote(self.project, safe="")
        _status, payload = self._send("GET", path, self._api_key, None)
        return payload

File: clients/ts/guardian-client.ts
/**
 * @fileoverview core-guardian client — vendor this single file into any
 * Cloudflare Worker. Zero runtime deps (fetch only). Identity comes from the
 * consumer's `GUARDIAN` object var; the two token audiences come from the
 * `GUARDIAN_AI_TOKEN` and `GUARDIAN_API_KEY` secret bindings. Source of truth:
 * https://github.com/jmbish04/core-guardian/blob/main/clients/ts/guardian-client.ts
 */

export type Priority = "hobby" | "normal" | "important" | "critical";
export type Importance = "low" | "medium" | "high";

export type GuardianConfig = {
  project: string;
  repo?: string;
  priority?: Priority;
  budget?: number;
  baseUrl?: string;
};

export type RunInput = {
  provider: string;
  model: string;
  input: unknown;
  importance?: Importance;
  mode?: "gateway" | "gateway-custom" | "provider-sdk-gateway" | "openai-compat" | "native" | "gemini-native";
  aiGatewayId?: string;
  transport?: "ai-sdk" | "provider-sdk" | "openai-compat" | "gemini-sdk";
  providerApiKey?: string;
};

export type RunResult = {
  request_uuid: string;
  status: number;
  provider: string;
  model: string;
  mode: string;
  gateway: string | null;
  tokens_in: number;
  tokens_out: number;
  cost_usd: number;
  body: unknown;
};

export type RegisterInput = {
  provider: string;
  model: string;
  tokensIn?: number;
  tokensOut?: number;
  tokensThinking?: number;
  requests?: number;
  costUsd?: number;
  operationId?: string;
  taskDescription?: string;
};

export type RegisterResult = {
  registrationId: string;
  id: string;
  day: string;
  worker: string;
  gateway: string;
  provider: string;
  model: string;
  requests: number;
  costUsd: number;
  tokensIn: number;
  tokensOut: number;
  tokensThinking: number;
  priced: "explicit" | "scraped" | "unmatched";
};

const DEFAULT_BASE_URL = "https://core-guardian.hacolby.workers.dev";
const PRIORITY_TO_IMPORTANCE: Record<Priority, Importance> = {
  hobby: "low",
  normal: "low",
  important: "medium",
  critical: "high",
};

export class GuardianError extends Error {
  status: number;
  body: unknown;
  isCircuitBreaker: boolean;
  circuitBrokenMessage?: string;
  constructor(status: number, body: unknown) {
    super(`Guardian request failed (${status})`);
    this.name = "GuardianError";
    this.status = status;
    this.body = body;
    const b = (body ?? null) as Record<string, unknown> | null;
    this.isCircuitBreaker = Boolean(b && b.isCircuitBreaker);
    if (b && typeof b.circuitBrokenMessage === "string") this.circuitBrokenMessage = b.circuitBrokenMessage;
  }
}

type Opts = GuardianConfig & { aiToken?: string; apiKey?: string; fetch?: typeof fetch };

async function safeJson(res: Response): Promise<unknown> {
  try {
    return await res.json();
  } catch {
    return null;
  }
}

export class GuardianClient {
  static readonly VERSION = "1.0.0";

  private cfg: GuardianConfig;
  private baseUrl: string;
  private fetchImpl: typeof fetch;
  // `declare` emits NO runtime field — the tokens are defined as non-enumerable
  // properties in the constructor, so neither JSON.stringify nor
  // util.inspect/console.log (Workers Logs) ever surfaces them.
  private declare readonly aiToken?: string;
  private declare readonly apiKey?: string;

  constructor(opts: Opts) {
    if (!opts.project) throw new Error("GuardianClient: config.project is required");
    // Narrowed copy — never spread aiToken/apiKey into cfg, or a logged/serialized
    // client instance would leak both secrets.
    this.cfg = { project: opts.project, repo: opts.repo, priority: opts.priority, budget: opts.budget, baseUrl: opts.baseUrl };
    this.baseUrl = (opts.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.fetchImpl = opts.fetch ?? fetch;
    // Non-enumerable so the tokens are excluded from every serialization path
    // (JSON.stringify, util.inspect, structured loggers) while staying readable
    // as this.aiToken / this.apiKey.
    Object.defineProperty(this, "aiToken", { value: opts.aiToken, enumerable: false });
    Object.defineProperty(this, "apiKey", { value: opts.apiKey, enumerable: false });
  }

  static fromEnv(env: Record<string, unknown>): GuardianClient {
    const raw = env.GUARDIAN;
    let cfg: GuardianConfig | undefined;
    try {
      cfg = (typeof raw === "string" ? JSON.parse(raw) : raw) as GuardianConfig | undefined;
    } catch {
      throw new Error("GuardianClient.fromEnv: env.GUARDIAN is not valid JSON");
    }
    if (!cfg || !cfg.project) throw new Error("GuardianClient.fromEnv: env.GUARDIAN.project missing");
    return new GuardianClient({
      ...cfg,
      aiToken: env.GUARDIAN_AI_TOKEN as string | undefined,
      apiKey: env.GUARDIAN_API_KEY as string | undefined,
    });
  }

  private importanceFor(over?: Importance): Importance {
    return over ?? (PRIORITY_TO_IMPORTANCE[this.cfg.priority ?? "normal"] ?? "low");
  }

  private runBody(i: RunInput, stream: boolean) {
    return {
      project: this.cfg.project,
      importance: this.importanceFor(i.importance),
      provider: i.provider,
      model: i.model,
      input: i.input,
      mode: i.mode,
      aiGatewayId: i.aiGatewayId,
      transport: i.transport,
      providerApiKey: i.providerApiKey,
      stream,
    };
  }

  private async post(path: string, token: string | undefined, body: unknown, raw = false): Promise<Response> {
    if (!token) throw new Error(`GuardianClient: missing token for ${path}`);
    const res = await this.fetchImpl(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: { "content-type": "application/json", authorization: `Bearer ${token}` },
      body: JSON.stringify(body),
    });
    if (!raw && !res.ok) throw new GuardianError(res.status, await safeJson(res));
    return res;
  }

  private async getJson(path: string, token: string | undefined): Promise<unknown> {
    if (!token) throw new Error(`GuardianClient: missing token for ${path}`);
    const res = await this.fetchImpl(`${this.baseUrl}${path}`, {
      headers: { authorization: `Bearer ${token}` },
    });
    if (!res.ok) throw new GuardianError(res.status, await safeJson(res));
    return res.json();
  }

  private async runImpl(i: RunInput): Promise<RunResult> {
    const res = await this.post("/api/ai-router/run", this.aiToken, this.runBody(i, false));
    return (await res.json()) as RunResult;
  }

  private async *streamImpl(i: RunInput): AsyncIterable<Uint8Array> {
    const res = await this.post("/api/ai-router/run", this.aiToken, this.runBody(i, true), true);
    if (!res.ok) throw new GuardianError(res.status, await safeJson(res));
    if (!res.body) return;
    const reader = res.body.getReader();
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      if (value) yield value;
    }
  }

  private async registerImpl(u: RegisterInput): Promise<RegisterResult> {
    const res = await this.post("/api/guardian/usage/register", this.apiKey, {
      worker: this.cfg.project,
      ...u,
    });
    return (await res.json()) as RegisterResult;
  }

  get ai() {
    return {
      run: (i: RunInput): Promise<RunResult> => this.runImpl(i),
      stream: (i: RunInput): AsyncIterable<Uint8Array> => this.streamImpl(i),
    };
  }

  get usage() {
    return {
      register: (u: RegisterInput): Promise<RegisterResult> => this.registerImpl(u),
    };
  }

  budget(): Promise<unknown> {
    return this.getJson("/api/ai/budget", this.apiKey);
  }

  project(): Promise<unknown> {
    return this.getJson(`/api/guardian/projects/${encodeURIComponent(this.cfg.project)}`, this.apiKey);
  }

  // The non-enumerable token properties already keep secrets out of
  // JSON.stringify; toJSON additionally drops the fetch impl for a tidy dump.
  toJSON(): unknown {
    return { cfg: this.cfg, baseUrl: this.baseUrl };
  }
}
